File: program/cubelist/apps/pc-version - backup/src/auth/pairing.rs
//! 페어링 QR 페이로드 처리
//!
//! 페이로드 = `{session_token, otp_secret, expires_at, device_fingerprint}`
//! TTL 60초, 사용 즉시 무효 (otp_secret은 1회용)
//!
//! Supabase Auth 세션 + QR 2-factor 조합:
//! 1. 모바일 PWA가 헬퍼에 페어링 요청 (Supabase Auth로 같은 계정 확인)
//! 2. 헬퍼가 QR 페이로드 생성 → 화면에 QR 표시
//! 3. 모바일이 QR 스캔 → otp_secret을 Supabase Edge Function에 검증 의뢰
//! 4. Edge Function이 user_devices 테이블에 등록 + 성공 응답
//! 5. 헬퍼는 otp_secret 즉시 폐기

use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};
use thiserror::Error;

use super::PAIRING_QR_TTL_SECONDS;

#[derive(Debug, Error)]
pub enum PairingTtlError {
    #[error("pairing payload expired (more than {0}s old)")]
    Expired(u64),

    #[error("pairing payload from the future")]
    FutureTimestamp,

    #[error("system time error: {0}")]
    SystemTime(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PairingPayload {
    /// Supabase Auth session token (헬퍼는 검증만, 발급은 PWA 측)
    pub session_token: String,

    /// 1회용 페어링 secret (Edge Function이 검증 후 즉시 폐기)
    pub otp_secret: String,

    /// UNIX timestamp seconds — TTL 60초 검증
    pub expires_at: u64,

    /// 페어링 대상 기기 fingerprint (PWA의 userAgent+canvas 해시 등)
    pub device_fingerprint: String,
}

impl PairingPayload {
    /// 페이로드의 TTL 검증 (현재 시각 기준)
    pub fn check_ttl(&self) -> Result<(), PairingTtlError> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|e| PairingTtlError::SystemTime(e.to_string()))?
            .as_secs();

        if self.expires_at < now {
            return Err(PairingTtlError::Expired(now - self.expires_at));
        }
        // 미래 시각은 max 5초 허용 (시계 오차)
        if self.expires_at > now + PAIRING_QR_TTL_SECONDS + 5 {
            return Err(PairingTtlError::FutureTimestamp);
        }
        Ok(())
    }

    /// 새 페이로드 생성 (헬퍼가 QR 표시 직전 호출)
    pub fn new(session_token: String, otp_secret: String, device_fingerprint: String) -> Self {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();
        Self {
            session_token,
            otp_secret,
            expires_at: now + PAIRING_QR_TTL_SECONDS,
            device_fingerprint,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new_payload_ttl_valid() {
        let p = PairingPayload::new("s".into(), "o".into(), "d".into());
        p.check_ttl().unwrap();
    }

    #[test]
    fn expired_payload_rejected() {
        let mut p = PairingPayload::new("s".into(), "o".into(), "d".into());
        p.expires_at = 0; // far in the past
        let err = p.check_ttl().unwrap_err();
        matches!(err, PairingTtlError::Expired(_));
    }

    #[test]
    fn future_payload_rejected() {
        let mut p = PairingPayload::new("s".into(), "o".into(), "d".into());
        p.expires_at = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs()
            + 99_999;
        let err = p.check_ttl().unwrap_err();
        matches!(err, PairingTtlError::FutureTimestamp);
    }
}

File: program/cubelist/apps/mobile-pwa/lib/pairing.ts
/**
 * 페어링 (모바일 PWA 측)
 *
 * 회사: 리버스 스테이션 (Rebirth Station)
 * 정착본: docs/tech-review.md §4, §6 (Rust auth / Security)
 *
 * 흐름
 * 1. 모바일 PWA가 Supabase Auth로 로그인 (같은 계정 보장)
 * 2. PC 앱 화면의 QR 스캔 → PairingPayload 획득
 * 3. PairingPayload.otp_secret을 Supabase Edge Function `/pair-device`로 전송
 * 4. Edge Function이 user_devices에 등록 + HMAC secret 안전 채널로 전달
 * 5. PWA는 HMAC secret을 IndexedDB에 보관, 매 WS 연결마다 HMAC nonce 생성
 *
 * **주의**: HMAC secret은 절대 localStorage에 저장 금지 (XSS 노출).
 * IndexedDB + SubtleCrypto 비추출 키 사용.
 */

export interface PairingPayload {
  session_token: string;
  otp_secret: string;
  /** UNIX timestamp seconds */
  expires_at: number;
  device_fingerprint: string;
}

export class PairingError extends Error {
  constructor(public readonly code: PairingErrorCode, message: string) {
    super(message);
    this.name = 'PairingError';
  }
}

export type PairingErrorCode =
  | 'qr_parse_failed'
  | 'expired'
  | 'future_timestamp'
  | 'edge_function_failed'
  | 'fingerprint_mismatch';

const QR_TTL_SECONDS = 60;
const CLOCK_SKEW_SECONDS = 5;

/**
 * QR 스캔 결과(JSON 문자열) → PairingPayload + TTL 검증
 */
export function parsePairingQr(raw: string): PairingPayload {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (e) {
    throw new PairingError('qr_parse_failed', (e as Error).message);
  }
  if (!isPairingPayload(parsed)) {
    throw new PairingError('qr_parse_failed', 'invalid pairing payload shape');
  }
  checkTtl(parsed);
  return parsed;
}

function isPairingPayload(v: unknown): v is PairingPayload {
  if (typeof v !== 'object' || v === null) return false;
  const o = v as Record<string, unknown>;
  return (
    typeof o.session_token === 'string' &&
    typeof o.otp_secret === 'string' &&
    typeof o.expires_at === 'number' &&
    typeof o.device_fingerprint === 'string'
  );
}

function checkTtl(p: PairingPayload): void {
  const now = Math.floor(Date.now() / 1000);
  if (p.expires_at < now) {
    throw new PairingError('expired', `payload expired ${now - p.expires_at}s ago`);
  }
  if (p.expires_at > now + QR_TTL_SECONDS + CLOCK_SKEW_SECONDS) {
    throw new PairingError('future_timestamp', 'payload from too far future');
  }
}

// ===========================================================================
// 기기 fingerprint (PWA 측)
// ===========================================================================

/**
 * 기기 fingerprint 생성 — userAgent + canvas 해시
 * 한 번 생성 후 IndexedDB에 보관, 같은 기기에서 재사용
 */
export async function generateDeviceFingerprint(): Promise<string> {
  const ua = navigator.userAgent;
  const canvas = renderFingerprintCanvas();
  const data = new TextEncoder().encode(ua + '|' + canvas);
  const digest = await crypto.subtle.digest('SHA-256', data);
  return hexEncode(new Uint8Array(digest));
}

function renderFingerprintCanvas(): string {
  // 캔버스가 없는 환경(SSR 등) 안전 처리
  if (typeof document === 'undefined') return 'no-canvas';
  const canvas = document.createElement('canvas');
  canvas.width = 200;
  canvas.height = 50;
  const ctx = canvas.getContext('2d');
  if (!ctx) return 'no-ctx';
  ctx.textBaseline = 'top';
  ctx.font = '14px Arial';
  ctx.fillText('큐브 리스트 fingerprint v1', 2, 2);
  ctx.fillStyle = 'rgba(102, 204, 0, 0.7)';
  ctx.fillRect(125, 1, 62, 20);
  return canvas.toDataURL();
}

function hexEncode(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

// ===========================================================================
// HMAC 생성 (WS 연결 시)
// ===========================================================================

/**
 * HMAC-SHA256(secret, nonce || timestamp_ms || device_id)
 *
 * Rust 측 verify_hmac()과 동일 입력 정의. 바이트 표현은 timestamp_ms를
 * **big-endian 8바이트** 로 인코딩 (Rust to_be_bytes()와 일치).
 */
export async function computeHmacHex(
  secret: ArrayBuffer,
  nonce: string,
  timestampMs: number,
  deviceId: string
): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    secret,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );

  // timestamp_ms를 big-endian 8바이트로 변환
  const tsBuf = new ArrayBuffer(8);
  const tsView = new DataView(tsBuf);
  // JavaScript number는 53-bit 정수 한계 → BigInt 사용
  tsView.setBigUint64(0, BigInt(timestampMs), false /* big-endian */);

  const encoder = new TextEncoder();
  const nonceBytes = encoder.encode(nonce);
  const deviceBytes = encoder.encode(deviceId);

  // 단일 ArrayBuffer로 결합 (subtle.sign이 한 번에 받음)
  const total = new Uint8Array(nonceBytes.length + 8 + deviceBytes.length);
  total.set(nonceBytes, 0);
  total.set(new Uint8Array(tsBuf), nonceBytes.length);
  total.set(deviceBytes, nonceBytes.length + 8);

  const sig = await crypto.subtle.sign('HMAC', key, total);
  return hexEncode(new Uint8Array(sig));
}

/**
 * nonce 생성 — 충분히 임의여야 함 (재사용 방지 캐시 우회 X)
 */
export function generateNonce(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return hexEncode(bytes);
}
